src/backend/distributed/transaction/distributed_deadlock_detection.c

```diff
@@ -24,12 +26,335 @@
+			Assert(list_length(deadlockPath) > 1);
+
+			LogDistributedDeadlockDebugMessage("Distributed deadlock found among the "
+											   "following distributed transactions:");
+
+			/*
+			 * We search for the youngest participant for two reasons
+			 * (i) predictable results (ii) cancel the youngest transaction
+			 * (i.e., if a DDL continues for 1 hour and deadlocks with a
+			 * SELECT continues for 10 msec, we prefer to cancel the SELECT).
+			 *
+			 * We're also searching for the youngest transactions initiated by
+			 * this node.
+			 */
+			foreach(participantTransactionCell, deadlockPath)
+			{
+				TransactionNode *currentNode =
+					(TransactionNode *) lfirst(participantTransactionCell);
+
+				TimestampTz youngestTimestamp =
+					youngestTransaction->transactionId.timestamp;
+				TimestampTz currentTimestamp = currentNode->transactionId.timestamp;
+
+				AssocateDistributedTransactionWithBackendProc(currentNode);
+
+				LogTransactionNode(currentNode);
+
+				if (currentNode->transactionId.initiatorNodeIdentifier ==
+					GetLocalGroupId() &&
+					timestamptz_cmp_internal(currentTimestamp, youngestTimestamp) == 1)
+				{
+					youngestTransaction = currentNode;
+				}
+			}
+
+			/* we should find the backend */
+			Assert(youngestTransaction->initiatorProc != NULL);
+
+			CancelTransactionDueToDeadlock(youngestTransaction->initiatorProc);
+			LogCancellingBackend(youngestTransaction);
+
+			hash_seq_term(&status);
+
+			return true;
+		}
+	}
+
+	LogDistributedDeadlockDebugMessage("No distributed deadlocks found");
+
+	return false;
+}
+
+
+/*
+ * CheckDeadlockForDistributedTransaction does a DFS starting with the given
+ * transaction node and checks for a cycle (i.e., the node can be reached again
+ * while traversing the graph).
+ *
+ * Finding a cycle  indicates a distributed deadlock and the function returns
+ * true on that case. Also, the deadlockPath is filled with the transaction
+ * nodes that form the cycle.
+ */
+static bool
+CheckDeadlockForTransactionNode(TransactionNode *startingTransactionNode,
+								TransactionNode **transactionNodeStack,
+								List **deadlockPath)
+{
+	List *waitingTransactionNodes = list_copy(startingTransactionNode->waitsFor);
+	int currentStackDepth = 0;
+	ListCell *waitingTransactionCell = NULL;
+
+	/*
+	 * We keep transactionNodeStack to keep track of the deadlock paths. At this point,
+	 * adjust the depth of the starting node and set the stack's first element with
+	 * the starting node.
+	 */
+	startingTransactionNode->currentStackDepth = currentStackDepth;
+	transactionNodeStack[currentStackDepth] = startingTransactionNode;
+
+	/* iterate on the children of the starting node and set their depths */
+	foreach(waitingTransactionCell, waitingTransactionNodes)
+	{
+		TransactionNode *waitingTransactionNode =
+			(TransactionNode *) lfirst(waitingTransactionCell);
+
+		waitingTransactionNode->currentStackDepth =
+			startingTransactionNode->currentStackDepth + 1;
+	}
```

Comment:
I don't think this is quite right. We shouldn't keep the stack depth in the TransactionNode, it should be part of the `waitingTransactionNodes` list (which I think would better be called to `todoList`, or `queudNodes` or something - everything is waiting here, so `waitingTransactionNodes` isn't very descriptive.

What I'm basically suggesting is that the "todo list" keeps not just a pointer to the node, but instead a pair ala `(pointerToNode, stackDepthWhenQueueing)`. Storing the depth inside the node isn't correct - a node can be queued from multiple places in the graph...